cabal-install/Distribution/Client/FileStatusCache.hs

```diff
@@ -0,0 +1,819 @@
+
+updateFileMonitor
+  :: (Binary a, Binary b)
+  => FileMonitor a b          -- ^ cache file path
+  -> FilePath                 -- ^ root directory
+  -> [MonitorFilePath]        -- ^ files of interest relative to root
+  -> a                        -- ^ the current key value
+  -> b                        -- ^ the current result value
+  -> IO ()
+updateFileMonitor monitor root monitorFiles cachedKey cachedResult = do
+    fsc <- buildMonitorStateFileSet root monitorFiles
+    rewriteCacheFile monitor fsc cachedKey cachedResult
+
+buildMonitorStateFileSet :: FilePath          -- ^ root directory
+                         -> [MonitorFilePath] -- ^ patterns of interest relative to root
+                         -> IO MonitorStateFileSet
+buildMonitorStateFileSet root =
+    go Map.empty []
+  where
+    go :: Map FilePath MonitorStateFile -> [FileGlobMonitorState]
+       -> [MonitorFilePath] -> IO MonitorStateFileSet
+    go !singlePaths !globPaths [] =
+      return (MonitorStateFileSet singlePaths globPaths)
+
+    go !singlePaths !globPaths (MonitorFile path : monitors) = do
+      let file = root </> path
+      monitorState <- handleDoesNotExist MonitorStateFileChanged $
+                        MonitorStateFile <$> getModificationTime file
+      let singlePaths' = Map.insert path monitorState singlePaths
+      go singlePaths' globPaths monitors
+
+    go !singlePaths !globPaths (MonitorFileHashed path : monitors) = do
+      let file = root </> path
+      monitorState <- handleDoesNotExist MonitorStateFileHashChanged $
+                        MonitorStateFileHashed
+                          <$> getModificationTime file
+                          <*> readFileHash file
+      let singlePaths' = Map.insert path monitorState singlePaths
+      go singlePaths' globPaths monitors
+
+    go !singlePaths !globPaths (MonitorNonExistentFile path : monitors) = do
+      let singlePaths' = Map.insert path MonitorStateFileNonExistent singlePaths
+      go singlePaths' globPaths monitors
+
+    go !singlePaths !globPaths (MonitorFileGlob globPath : monitors) = do
+      monitorState <- buildFileGlobMonitorState root "." globPath
+      go singlePaths (monitorState : globPaths) monitors
+
+
+buildFileGlobMonitorState :: FilePath     -- ^ the root directory
+                          -> FilePath     -- ^ directory we are examining relative to the root
+                          -> FilePathGlob -- ^ the matching glob
+                          -> IO FileGlobMonitorState
+buildFileGlobMonitorState root dir globPath = do
+    dirEntries <- getDirectoryContents (root </> dir)
+    dirMTime   <- getModificationTime (root </> dir)
+    case globPath of
+      GlobDir glob globPath' -> do
+        subdirs <- filterM (\subdir -> doesDirectoryExist (root </> dir </> subdir))
+                 $ filter (globMatches glob) dirEntries
+        subdirStates <-
+          forM subdirs $ \subdir -> do
```

Comment:
Shouldn't this be `forM (sort subdirs)` (as in the `GlobFile` case) to preserve the invariant?